### ISSUE TITLE
Suboptimal additional indentation for decorated functions

### ISSUE THREAD

**Issue by mcjazzyfunky** [prettier/prettier#10709]
**Prettier 2.2.1**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEB6VACAvDjcAeAhgLYAOANnBgIzY70ONPMtMYA6U6GMAFlQDMIAJ2KEY8ACYYIAVxil5MgT14BLAM4yoVAO5ry5beQCeGchAgBrDAHNLkztzUrSwuBLVxhGAG4AmADog2k0MWQ04aQxCLTUYHndxLQFZKDAYNWgtPnEnTEJ3DCgSKI4QSTgBNR12EAxJFwFvBATcqBk+by1JWSoYCAwAI1lbDABiagAGABZ-fwwAWnyMSKpeCVINJHRbeN5ZIcDIYlQ3D0zvM-dPK80NPo1Uabn-Tk5IKA0EgGE5WG82AaVRqcAAFAByYgmRaQNLwYQQgA0GDBwE4GAwNXiakI5D+8KBUyRGPMhCGcCMWAwEIJAMRnAAvkgMHSEQAFYQQLYASmwAD4MOiOhhPt8MABtOGwFGRGBsgC6QIicAAyjBxODsZk8WyeaSxQloD9yGowDZqSqfnjyENCOawWC+VhBXK2WD4gKsQkANQ0HkoiUK-Ui0k3WTCDpg0mYgA8Iwk0G0YFN5qwwGNqasjP5McxQvI5MpzKF0pgjLzsdQCYGUFzIr5mKZIZW9DwRDIlAwC1Yvb7-Y4UBWi0FXRikkamSTNUqsHEWQ6NVUghEYgkZTkCnkK3zYXJGgg5HklLMUAgCV0hAB0n47ne6Wyv3+CKBJ1I0FakOhsOf3mRqOFTFtVxfFnyJEkRULCkqRpNk-yZFk4OETluQ0Z1BUA0VH0lMtZQ8RVlUidVNQ9KAcV1Z8Q0xQ1tBNM0LXCSJrUMO0HSdL03WfD0Ehdb0MD9agA0lYNSTDDwIyjSsayTaAU3o9NM3onM80xYAoOLFlgDLCsRTjat5FretMUbQdGRDEAkRAbkpy+ZBQEKLldHZQoEA0ZAQDxS8TDcyyhmEe0rA8VVSHtGpbGQGBhD6Sy4GICkJyiAAZK9bFkQhbDgAAxVdxEyKBwpQQh5AgCyQA2YhyAAdXUeANBCsA1VcnFfHiEx3LADQfJAGpImEGBOXSsRkAEPFIksgArDR8AAIX880gtKRLQWG0a4AmqbVTCygAEVZHPOAVvIMaQBC4RevcmATFIOANDAYQ1FIGBSrcGoYEqtRJD4ZAAA5iROrlIkq-zSHc85et8A7LIARz2+AUK2dzYkWHQoiiUr3BhtR3AG2whqQEajrWkBImINRDuOjQtrgXb9oiqKiY1IZ3s+3hkH8SzIsIAwwr+Yg8ZAM8dFKlUABV93JonfD6ABJKBZxgVU7oemAAEE5fVExKAlxlGSAA)

```sh
--parser typescript
--no-semi
--single-quote
--trailing-comma none
```

**Input:**

```tsx
// === example 1 ========================================== 
// the formatted output of this one will only look good
// if prettier v2.2.1 is used  as it treats functions that
// are named "define" different than others due to bug #10422 -
// see https://github.com/prettier/prettier/issues/10422

const Counter = define('my-counter', ({
  initialCount = 0,
  label = 'Counter'
}: CounterProps) => {
  const [count, setCount] = useState(initialCount)
  const onClick = useCallback(() => setCount(it => it + 1), [])
  
  return (
    <button onclick={onClick}>
      {label}: {count}
    </button>
  )   
})

// === example 2 ========================================== 

// -> the addition indentation in the formatted output
//    is absolutely not wanted here

const Counter = component('my-counter', ({
  initialCount = 0,
  label = 'Counter'
}: CounterProps) => {
  const [count, setCount] = useState(initialCount)
  const onClick = useCallback(() => setCount(it => it + 1), [])
  
  return (
    <button onclick={onClick}>
      {label}: {count}
    </button>
  )   
})

```

**Output:**

```tsx
// === example 1 ==========================================
// the formatted output of this one will only look good
// if prettier v2.2.1 is used  as it treats functions that
// are named "define" different than others due to bug #10422 -
// see https://github.com/prettier/prettier/issues/10422

const Counter = define('my-counter', ({
  initialCount = 0,
  label = 'Counter'
}: CounterProps) => {
  const [count, setCount] = useState(initialCount)
  const onClick = useCallback(() => setCount((it) => it + 1), [])

  return (
    <button onclick={onClick}>
      {label}: {count}
    </button>
  )
})

// === example 2 ==========================================

// -> the addition indentation in the formatted output
//    is absolutely not wanted here

const Counter = component(
  'my-counter',
  ({ initialCount = 0, label = 'Counter' }: CounterProps) => {
    const [count, setCount] = useState(initialCount)
    const onClick = useCallback(() => setCount((it) => it + 1), [])

    return (
      <button onclick={onClick}>
        {label}: {count}
      </button>
    )
  }
)

```

**Expected behavior:**
Would be perfect if the function indentation stays as is (no additional two space indentation) - the  formatting of the argument list is not that important.
Please ignore that the examples above look a bit like React or Preact - it's not about those libraries, it's just about the pattern.
Very often I have the case that special functions have the purpose to transform other functions, often including a bit of meta information (like a string or whatever), like in the examples above, but then the formatting by prettier will be quite ugly.
You may say for the examples above just define first a named function and then call the `component` function afterwards with the function variable as argument ... but in general this may explicitly not wanted and be more complicated (for example if you have `const Counter = component('my-counter', counterOptions, func)` where the type of `func` depends heavily on the type of `counterOptions`). Of course with some aditional lines of code and some utility types this is doable, but will be syntactically ugly and less readable.
Frankly, I have no concrete suggestion what could be a good solution regarding what format should be applied in what concrete cases - but I hoped you may have some ideas.

[Edit] One possibility could be: If all arguments except for the last one are literals or variables and the last argument is a closure and the literals and variables match one line then format as suggested above otherwise format as usual.

**Comment by thorn0** [comment]
Closely related: https://github.com/prettier/prettier/issues/6921

**Comment by thorn0** [comment]
I don't believe there is a solution. In situations like this, the developer is willing to sacrifice the readability of a call to get less indentation. How can an algorithm guess this intention?

For `it`, `describe`, and co this was solved by hardcoding well-known function names, but that obviously doesn't scale.

Moreover, I suspect many people from #6921 would say that the current formatting is good because this is the best formatting for function composition.

Let's keep the issue open for discussion, but don't really expect this to lead anywhere.

**Comment by thorn0** [comment]
Actually, there exists a way to refactor your code to get the formatting you want, but 1) you're most probably not going to like it, and 2) I'm not sure how much intended it is (see #9662):

**Prettier 2.2.1**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEACAOlSUDONUDCEArrHAE6oC8qkAtgA7QIwAUA5HQJ4C0kp8cuwCUrAGakwMAJbRWwTKlTSo0mQEMANkQHVUABgA0i1JvUAjOJr3sdZIZgC+SQiXsAFchAY5hS1ApQSth4qADa-LCGqDhwMHYwALp6xLEAyjDq8KwqatJaCcImIfjQBJrSYADWKbEEWprm6tWsrH5UAHwxcQk5+J3K+ADUqACMwtFhiUVBGLPkccTkQawmSgA85sQwMNCo0GAV1VTAZUdVjh1r-sBmlprOAZEwjtfrAPRbO9BXs35KThmIEMIG8MmgOGQoHU5C8AHd3DCEJCUFo4eouJCQeZyM0qnE0gxmioAObIGDkYhwEFwOiWAAm9Lg9IAMuooCTiOoSXAAGIQch0LIyDnIEDqbYQYEgAAWMDomgA6jK1HAcESwHA0si8gA3NRcMVgHBYkAqWLkGCeblC5BiLSxEEAKxwAA8AEK46oE9R0OAslRwO0O6kgF2utKkzRwACKxAg8GDmkdICJ5AtYpgXAYarA5GkDBg0oY+dgiuk9JgMuQAA4jKmvLFFbiGGKS2qKLqgyCAI7x+Cebwo8U4HhQODM5nShZ96QLa0k21Ie3J0OxOjSJMpnBR2P9oNIClUkGZczlyvVpAAJhPuOkFQ5RDoS5AUGY0tScAAKhYUSuU7qVIAJJQEysBpHmBYwAAgqBGRcNGW5wI4jhAA)

**Input:**

```tsx
 
const Counter = component('my-counter')(function({
  initialCount = 0,
  label = 'Counter'
}: CounterProps)   {
  const [count, setCount] = useState(initialCount)
  const onClick = useCallback(() => setCount(it => it + 1), [])
  
  return (
    <button onclick={onClick}>
      {label}: {count}
    </button>
  )   
})

```

**Output:**

```tsx
const Counter = component('my-counter')(function ({
  initialCount = 0,
  label = 'Counter'
}: CounterProps) {
  const [count, setCount] = useState(initialCount)
  const onClick = useCallback(() => setCount((it) => it + 1), [])

  return (
    <button onclick={onClick}>
      {label}: {count}
    </button>
  )
})

```

**Comment by js-works** [comment]
@thorn0 Thanks a lot for the info. Indeed I knew that curring the above function would help. But it feels a bit odd to curry a function or make it partially applicable in general just to suit one of your development tools. But at least it is some kind of workaround.

**Comment by mcjazzyfunky** [comment]
@thorn0 Thanks a lot for the info. Indeed I knew that curring the above function would help. But it feels a bit odd to curry a function or make it partially applicable in general just to suit one of your development tools. But at least it is some kind of workaround.

**Comment by thorn0** [comment]
It's currying + using `function` instead of `=>`. Odd, indeed.

**Comment by thorn0** [comment]
If function decorators were a thing, they'd work similarly (with partial application), wouldn't they? So I think we're onto a pattern here. We can try supporting this pattern for arrow functions too:

```
A(B)(C)
where
  A - identifier
  B - 0 or more call arguments
  C - an arrow function, the only argument of the outer call
```
Proposal: in this case and only in it, the arrow function `C` can be kept hugged even if its signature breaks.

**Comment by mcjazzyfunky** [comment]
In fact the other day I've experimented a bit with the following syntax (this is semantically different form the example above, but like said it's just about the patterns):

**Prettier 2.2.1**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuc0DOMAEBhCBXWOAJ0wF5MATOAMwEso4AKAcgFsBPAWkgPiOYCUjRgAciEEWiSZgAHSiZM9WjFoBDADa5eAfmlQ8rAEbEANPMUa1JjXswYi9AObyAvgLIA+GRcyQoGJgiZJh4aHAAIjRqeBowouKSpj4KikpQKupa+LDSAAzmqZbWcBrSzNqE-L7uvr7+gQDaaMnhMAC6IWFwAMowavCMwH45MNIiAHTKqpqVWLWpDVjQ2Bq0YADWIYwepN5tLDywzMkqXkpYANSYAIwCdalEcDB4RAo754y+igA8RngwGDQTDQMBrTakYArcEbVyeb5pYCTKw2VzSYBoCZHGCuBE-AD0-0B0HhqXuUAWIFMIAkqnQyFAaiI4gA7gAFJkINDIECaFlqdjc6lGIhqTbPHoiMXOZAwIh4ODUuDGOAUKgUAAyaigTjwaiccAAYhAiKwBqodTyYkCqSAABYwVgaADqdpUcDQUrAvS5mQAbip2DywGghSB6OEiDA2aKnGbkNRNOFqQArNAADwAQqLxTAempWHANfQ4Amk4qQGn0z1nBo4ABFPAQeBljTJkBSoiRnkwdgiD1gRwiGC2sT0GDO2gUGB25AADgKHfE4WdopEPLEHuIftL1IAjk34DGJNyUGo0JwGKrVbangfaE8Y-r40hE22K+FWLRZfKP7WG4epavuW1L9EYk7TrOSAAEygaKtBrDquCsC+IAegArLa3QACrWKeb7tn6CoAJJQFQsA9IOtDDgAgmRfTsHWrbhK4rhAA)

```sh
--parser typescript
```

**Input:**

```tsx
const Counter = define('my-counter')((props: {
  initialCount?: number,
  label?: string
}) => {
  const p = useDefault(props, {
    initialCount: 0,
    label: 'Counter'
  })
  
  const [s, set] = useState({ count: p.initialCount })
  const onClick = () => set('count', it => it + 1)
  
  return () => (
    <button onclick={onClick}>
      {p.label}: {s.count}
    </button>
  )
})

```

**Output:**

```tsx
const Counter = define("my-counter")(
  (props: { initialCount?: number; label?: string }) => {
    const p = useDefault(props, {
      initialCount: 0,
      label: "Counter",
    });

    const [s, set] = useState({ count: p.initialCount });
    const onClick = () => set("count", (it) => it + 1);

    return () => (
      <button onclick={onClick}>
        {p.label}: {s.count}
      </button>
    );
  }
);

```

I've abondonned it as I didnt't like the format prettier is converting it in and using the `function` notation looked ugly.
This is an simple form of one of the patterns that you've mentioned above.
Yes, if the `=>` notation would behave like the `function` notation that could be at least a litte win (still am not that happy about currying here, but maybe writing `)(` instead of `, ` should not be such a big deal).

**Comment by thorn0** [comment]
It's quite possible we'll add support for this pattern unless our testing shows that it doesn't result in too many unwanted changes, which seems unlikely.

**Comment by thorn0** [comment]
> at least a litte win

Believe me, if at least some compromise can be reached in an issue like this, it's a huge win.

**Comment by leimonio** [comment]
@thorn0 Is there a way to override this behaviour. Upgrading to prettier 2.3.0 and above leads to so many changes in the codebase. It would be great if we could skip how this extra indentation is added.

**Comment by xdev1** [comment]
@thorn0 May I ask whether there are any news about this issue?

**Comment by thorn0** [comment]
@xdev1 The best that can be done here is to finalize #10714 (an implementation of this proposal: https://github.com/prettier/prettier/issues/10709#issuecomment-821065674). If you're asking about the input from the first message here (without currying), the formatting for it can't be changed. That would conflict with too many other expectations.

**Comment by xdev1** [comment]
@thorn0 Yes, the "currying" version (aka. "decorated function" version) is fine. I understand that the original request cannot be fulfilled.

The syntax in [this demo](https://deploy-preview-10714--prettier.netlify.app/playground/#N4Igxg9gdgLgprEAuc0DOMAEBhCBXWOAJ0wF5NIBbAB2gRgAoAdEXA+IlgSgYeqUzAmUTJigBDSnAEYiASygBzANzDRCuTDniANm1gB+AVDyUARsVVQAvlzIA+QWorosAbTQAaTGjgwAumQ+MOLwDMAu7ALUAHQaWrr6WAA+yZgADJi2VqKQUBiYCmBEcFKwQQx2pI6+MDGQ7AyaDoVYANSYAIxcVs4lMHhEIpUtzCKimAA8ZngwMNCY0Ng6cmAA1qTARSVlMNb2zhOCsRJS1gLAaPX4sNaHUwD0M3PQB+M9wtkgniAQ1FroZCgcREIgQADuAAUQQg0MgQLpweIAJ5wn5mIjidZ+ADK1CxCkUyBgRDwcB+pQsABMqXAqQAZcRKPDiRRwABiECIlFCWiU8PEswg3xAAAsYJQdAB1UWaOBofFgOA42GaOQAN00yPhYDQaJACl8RBgkMxih5yAAZrpfD8AFZoAAeACFMdiYDjJHB6Qo4FabeSQA7HTjCTo4ABFPAQeD+nS2kD4ohG+FmcQWHQAWhg+uo8lgUrkVJgouQAA50j88xBfFLMdR4Xn5cR1X6fgBHaPwU1-OEocRoTNQOB0ukikqduQlU2si1Ia3xwO+ShyYmkpdhyNdv3zgM-EJmQvF0tIABM+8xchWSlwlDnIHlAFYRXhfAAVdN9hcJ9VkgCSUC0rAOLFHI-wAIKATiMDIuGca+NY1hAA) (using PR #10714) looks like a good compromise IMHO and may be useful in other cases as well, I guess.

PS: Oh, btw: I'm the author of this issue, just using a different username nowadays.

**Comment by thorn0** [comment]
fixed by #10714
